File: 113_PathSum_II.py
###83.91%

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        def pathSumHelper(root, sum):
            if not root:
                return
            curr.append(root.val)
            if root.val == sum and not root.left and not root.right:
                res.append(curr[:])
                curr.pop()
                return
            pathSumHelper(root.left, sum - root.val)
            pathSumHelper(root.right, sum - root.val)
            curr.pop()
        res = list()
        curr = list()
        pathSumHelper(root, sum)
        return res
		
		
class Solution:

def pathSum(self, root, sum):
    if not root: return []
    if root.left == None and root.right == None:
        if sum == root.val: 
            return \[\[root.val\]\]
        else: 
            return []
    a = self.pathSum(root.left, sum - root.val) + \
        self.pathSum(root.right, sum - root.val)
    return [[root.val] + i for i in a]

File: 113_PathSum_II.cpp
//79.96%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(!root)
			return vector<vector<int> > ();
		vector<vector<int> > res;
		vector<int> curr;
		pathSumHelper(res, curr, root, sum);
		return res;
    }
	
private:
	void pathSumHelper(vector<vector<int> >& res, vector<int>& curr, TreeNode *root, int sum){
		if(!root)
			return;
		curr.push_back(root -> val);
		if(root -> val == sum && !root -> left && !root -> right){
			res.push_back(curr);
			curr.pop_back();
			return;
		}
		
		pathSumHelper(res, curr, root -> left, sum - root -> val);
		pathSumHelper(res, curr, root -> right, sum - root -> val);
		curr.pop_back();
	}
};